### DISCUSSION TITLE
Starting a database transaction in a mutation contesxt

### DISCUSSION THREAD

**Discussion by francipvb** [strawberry-graphql/strawberry#3380] [Q&A] [answered]
Hello all,

I am trying to ensure all mutations run inside a single database transaction.

I realyzed I need these things:

- A way to detect if the running query is a mutation, query or subscription.
- A way to run code before the the actual query resolvers.
- A way to run code after the actual query resolver(s).

I am using the FastAPI integration to leverage dependency injection, but I don't have a way to run some code before the query is executed and then another code chunk after the query execution ends, successfully or not.

The motivation for this is that I am developing a mutation system that adds modifying resolvers inside resolved mutation results got from root.

**Comment by DoctorJohn** [comment, marked as answer]
Hi, did you already take a look at the Strawberry extensions system?

You could possibly make use of the `on_operation` or `on_execute` lifecycle hooks to run code before and after execution of your resolvers. Via `self.execution_context` you should be able to figure out whether your operation is a query, mutation or subscription.

**Reply by francipvb** [reply]
Hello,

Yes, this is what I did.

**Reply by marmor157** [reply]
@francipvb Would you be willing to share your solution? I'm trying to do the same and I keep running into the issue about `sync_to_async` and `You cannot call this from an async_context` related stuff.

**Reply by patrick91** [reply]
@marmor157 what are you trying to do? are you using django?

**Reply by francipvb** [reply]
Hello @marmor157 

Use a strawberry extension to start and commit your transaction based on the operation type.

Just implement your `on_operation` as an async method and you'll be fine.

```python
# ... imports

class AsyncSessionExtension(SchemaExtension):

    async def on_operation(self):
        async with self.execution_context.context.session.begin():
            yield
```

This is a simple approach and your context must have the session factory somehow, or you can inject it from here.

**Reply by marmor157** [reply]
Thanks for the answer. 

Yeah, I forgot to mention that I'm using django, so sadly doing it in `on_execute` or `on_operation` like this:
```python
    async def on_execute(self) -> AsyncIteratorOrIterator[None]:
        if self.execution_context.operation_type == OperationType.MUTATION:
            async with transaction.atomic():
                yield
        else:
            yield
```
results in the `TypeError: 'Atomic' object does not support the asynchronous context manager protocol`. 

After a bit of tinkering I've came up with this
```python
class MutationTransactionExtension(SchemaExtension):
    async def resolve(self, next_, root, info, *args, **kwargs):
        if info.operation.operation == OperationType.MUTATION:

            @sync_to_async
            def run_in_transaction():
                with transaction.atomic():
                    return next_(root, info, *args, **kwargs)

            result = await run_in_transaction()
        else:
            result = next_(root, info, *args, **kwargs)
        # We have to await here, because return type from `run_in_transaction` can be coroutine, since we can't await it there
        return await await_maybe(result)
```

**Reply by francipvb** [reply]
Keep an eye on your code. I suggest you to avoid async operations because the AsyncIO support is not yet mature and probably you'll have issues related to queries and transactions.
